### PR TITLE
Eagerly propagate GOAWAY frames from the Http2FrameCodec

### LINKED ISSUE #9986: DefaultHttp2ConnectionDecoder triggers stream channel closes more eagerly than it should
### Expected behavior
When a Http2FrameCodec receives a GOAWAY the stream channels with ID's greater than the last seen will get notified of the GOAWAY so they can bubble up a retryable signal.

### Actual behavior
The connection is notified first which eagerly closes the streams. This happens [here](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L215).

A work around is to add a listener to the `Http2Connection` that eagerly sends another copy of the `Http2GoAwayFrame` down the pipeline. This works because [connection listeners get called](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2Connection.java#L234) before the extra streams are culled, but that is also painful because the `DefaultHttp2GoAwayFrame` [constructor](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2GoAwayFrame.java#L80) that allows your to set the `lastStreamId` is package private.

### Netty version
4.1.43.Final but also appears to affect 4.1

### PR DESCRIPTION
**Pull request by bryce-anderson**
Motivation:

Users of the Http2FrameCodec are notified of inbound GoAwayFrames
after the connection has already closed any ignored streams, potentially
losing the signal that some streams may have been ignored by the peer and
are thus retryable.

Modifications:

The Http2FrameCodec more eagerly sends the Http2GoAwayFrame from the
ConnectionListener which happens before the connection closes ignored
streams.

Result:
Fixes #9986

### CODE REVIEW

**Comment by bryce-anderson** [comment]
I'm not confident this is what we want since it's a bit of a hack. Another solution would be to simply reorder the connection and frame listener invocations in the `DefaultHttp2ConnectionDecoder` (found [here](https://github.com/netty/netty/blob/4.1/codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionDecoder.java#L217))  but that causes [this](https://github.com/netty/netty/blob/4.1/codec-http2/src/test/java/io/netty/handler/codec/http2/Http2ConnectionRoundtripTest.java#L948) test to fail. I think there is a reasonable argument to be made the the test is hoping for best case outcomes from pathological implementations (it's hard to imagine wanting to send headers based on receiving a goaway frame) and the regression is still correct: the stream may write it's headers but will still be given a goaway and closed.

**Comment by normanmaurer** [comment]
@bryce-anderson imho re-ordering in the `DefaultHttp2ConnectionDecoder` may be the better bet.

File: codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java [hunk 1]
```diff
@@ -348,6 +348,35 @@ public void receiveGoaway() throws Exception {
...
+        final AtomicBoolean wasStillOpen = new AtomicBoolean();
+
+        channel.pipeline().remove(inboundHandler);
+        inboundHandler = new LastInboundHandler() {
+            @Override
+            public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+                if (msg instanceof Http2GoAwayFrame) {
+                    wasStillOpen.set(frameCodec.connection().stream(4).state() == State.OPEN);
+                }
+                    super.channelRead(ctx, msg);
```
Comment (normanmaurer) [inline codec-http2/src/test/java/io/netty/handler/codec/http2/Http2FrameCodecTest.java]: nit: remove extra 4 spaces

**Comment by bryce-anderson** [comment]
> @bryce-anderson imho re-ordering in the `DefaultHttp2ConnectionDecoder` may be the better bet.

I tend to agree. I'll make a second PR for that one and we can decided with them both at hand.

**Comment by bryce-anderson** [comment]
See https://github.com/netty/netty/pull/10009 for the alternative, which I like better.

**Comment by bryce-anderson** [comment]
#10009 seems like the way forward. We can reopen this if we change our minds.

**Pull request closed without merging**